applications/CompressiblePotentialFlowApplication/custom_elements/embedded_compressible_potential_flow_element.cpp

```diff
@@ -123,6 +153,65 @@ void EmbeddedCompressiblePotentialFlowElement<Dim, NumNodes>::CalculateEmbeddedL
     noalias(rRightHandSideVector) = -prod(laplacian, potential);
 }
 
+template <int Dim, int NumNodes>
+void EmbeddedCompressiblePotentialFlowElement<Dim, NumNodes>::CalculateKuttaWakeLocalSystem(
+    MatrixType& rLeftHandSideMatrix, VectorType& rRightHandSideVector, const ProcessInfo& rCurrentProcessInfo)
+{
+    // Note that the lhs and rhs have double the size
+    if (rLeftHandSideMatrix.size1() != 2 * NumNodes ||
+        rLeftHandSideMatrix.size2() != 2 * NumNodes)
+        rLeftHandSideMatrix.resize(2 * NumNodes, 2 * NumNodes, false);
+    if (rRightHandSideVector.size() != 2 * NumNodes)
+        rRightHandSideVector.resize(2 * NumNodes, false);
+    rLeftHandSideMatrix.clear();
+    rRightHandSideVector.clear();
+
+    MatrixType laplacian_matrix = ZeroMatrix(2 * NumNodes, 2 * NumNodes);
+
+    PotentialFlowUtilities::ElementalData<NumNodes,Dim> data;
+
+    // Calculate shape functions
+    GeometryUtils::CalculateGeometryData(this->GetGeometry(), data.DN_DX, data.N, data.vol);
+    data.distances = PotentialFlowUtilities::GetWakeDistances<Dim, NumNodes>(*this);
+
+    const double density = BaseType::ComputeDensity(rCurrentProcessInfo);
+    const double DrhoDu2 = BaseType::ComputeDensityDerivative(density, rCurrentProcessInfo);
+
+    // Computing local velocity
+    array_1d<double, Dim> v_upper = PotentialFlowUtilities::ComputeVelocityUpperWakeElement<Dim, NumNodes>(*this);
+    array_1d<double, Dim> v_lower = PotentialFlowUtilities::ComputeVelocityLowerWakeElement<Dim, NumNodes>(*this);
+
+    const BoundedVector<double, NumNodes> DNV_upper = prod(data.DN_DX, v_upper);
+    const BoundedVector<double, NumNodes> DNV_lower = prod(data.DN_DX, v_lower);
+
+    const BoundedMatrix<double, NumNodes, NumNodes> laplacian_total =
+        data.vol * density * prod(data.DN_DX, trans(data.DN_DX));
+
+    const BoundedMatrix<double, NumNodes, NumNodes> lhs_total_upper =
+        data.vol * density * prod(data.DN_DX, trans(data.DN_DX)) +
+        data.vol * 2 * DrhoDu2 * outer_prod(DNV_upper, trans(DNV_upper));
+
+    const BoundedMatrix<double, NumNodes, NumNodes> lhs_total_lower =
+        data.vol * density * prod(data.DN_DX, trans(data.DN_DX)) +
+        data.vol * 2 * DrhoDu2 * outer_prod(DNV_lower, trans(DNV_lower));
+
+    for (unsigned int i = 0; i < NumNodes; ++i)
+    {
+        for (unsigned int j = 0; j < NumNodes; ++j)
+        {
+            rLeftHandSideMatrix(i, j) = lhs_total_upper(i, j);
+            rLeftHandSideMatrix(i + NumNodes, j + NumNodes) = lhs_total_upper(i, j);
+
+            laplacian_matrix(i, j) = laplacian_total(i, j);
+            laplacian_matrix(i + NumNodes, j + NumNodes) = laplacian_total(i, j);
+        }
+    }
+
+    BoundedVector<double, 2*NumNodes> split_element_values;
+    split_element_values = PotentialFlowUtilities::GetPotentialOnWakeElement<Dim, NumNodes>(*this, data.distances);
+    noalias(rRightHandSideVector) = -prod(laplacian_matrix, split_element_values);
```

Comment:
Short question: I understand you compute RHS and LHS together because it is much faster at the time of building. Don't we compute also the RHS after the solution to check for convergence? And if we do, does it pay off to compute the LHS again even if we don't need it?

I am referring to:

https://github.com/KratosMultiphysics/Kratos/blob/32400a493257fb8bdce93479f9e04eeed080c6df/kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h#L844

We discussed this at some point but it is not clear to me.

Reply:
take a look at the stuctural elements, esp the `CalculateAll` fct:
e.g. here: https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp#L315-L352

Reply:
Exactly that is what I meant. In [there](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp#L315-L352) `CalculateAll` is called each time `CalculateRightHandSide` is called, doesn't this create an unnecessary overhead (both in cost and memory)? Why compute the full LHS matrix when only the RHS vector is needed?

Reply:
I just saw the calculation flags, I think that might be the solution, but right now they seem to be hard coded and not used.

Reply:
Now I see that for the solid element it is done correctly: [here](https://github.com/KratosMultiphysics/Kratos/blob/b3fbc9d830c3fff5ca356ce24cad6535d425a680/applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp#L559-L574) and [here](https://github.com/KratosMultiphysics/Kratos/blob/b3fbc9d830c3fff5ca356ce24cad6535d425a680/applications/StructuralMechanicsApplication/custom_elements/total_lagrangian.cpp#L156-L245). Thanks for the hint @philbucher! 

@marcnunezc I think this approach could speed up the nonlinear computations in 3d, we can think of doing this in the potential flow elements at some point, it shouldn't be much work.

Reply:
> Short question: I understand you compute RHS and LHS together because it is much faster at the time of building. Don't we compute also the RHS after the solution to check for convergence? And if we do, does it pay off to compute the LHS again even if we don't need it?
> 
> I am referring to:
> 
> https://github.com/KratosMultiphysics/Kratos/blob/32400a493257fb8bdce93479f9e04eeed080c6df/kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h#L844
> 
> We discussed this at some point but it is not clear to me.

I see your concern, but I am not 100% convinced. There are also occasions were the full system is built right?
https://github.com/KratosMultiphysics/Kratos/blob/32400a493257fb8bdce93479f9e04eeed080c6df/kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h#L930
There are calculations that are dependent one on the other, so Im not sure what yields the minium number of calculations. 
https://github.com/KratosMultiphysics/Kratos/blob/a8a260e110a1534ca834623341f17deac802439c/applications/CompressiblePotentialFlowApplication/custom_elements/embedded_compressible_potential_flow_element.cpp#L142-L153

For instance here we are using the same loop to compute the LHS and RHS. Consider we build the full system once and then once again the RHS to check the convergence. This loop would be perfomred twice. If the LHS/RHS is separated, this loop would be perofrmed three times (one for the LHS and twice for the RHS). 

I am not sure what is the best option

Reply:
I see your point. These dependency of the calculations makes it difficult to assess what would be faster. I see your point with the number of loops too, but at the same time separating would make both loops also smaller (some calculations are not needed for the LHS and some are not needed for the RHS). Still, I'm also not convince, I guess the only way of knowing is to try both options in a big example. Also, I don't know how important this is, we would need to see how much time takes building compared to solving, meaning that there might be a difference between both approaches but it might be negligible compared to the time that takes to solve the system. 

Anyways, it was just a question I had, I don't think this should be a blocker nor a priority right now.

Reply:
:+1: 

This is something we could revisit when considering the refactoring of all elements, when creating a "potential_flow_base_element"

Reply:
Sounds good :+1: